src/hal/drivers/mesa-hostmot2/hm2_rpspi.c

```diff
@@ -0,0 +1,568 @@
+	uint8_t spibpf;	//bits per frame
+};
+
+static hm2_rpspi_t boards[MAX_BOARDS];
+static int comp_id;
+
+
+/*************************************************/
+//so untested, you could implement rtapi_open_as_root in "C" as
+int rtapi_open_as_root(const char *filename, int mode) {
+		setfsuid(geteuid());
+	int r = open(filename, mode);
+	setfsuid(getuid());
+	return r;
+}
+
+/*************************************************/
+// aib (address increment bit)
+static uint32_t mk_read_cmd(uint16_t addr, unsigned msglen, bool aib) {
+    return 0 | (addr << 16) | 0xA000 | (aib ? 0x800 : 0) | (msglen << 4);
+}
+
+/*************************************************/
+static uint32_t mk_write_cmd(uint16_t addr, unsigned msglen, bool aib) {
+    return 0 | (addr << 16) | 0xB000 | (aib ? 0x800 : 0) | (msglen << 4);
+}
+
+/*************************************************/
+/*
+SPI_CS_RXF		0x00100000	
+	RXF - RX FIFO Full
+		0 = RXFIFO is not full.
+		1 = RX FIFO is full. No further serial data will be sent/ received until data is read from FIFO.
+
+SPI_CS_RXR		0x00080000
+	RXR RX FIFO needs Reading ( full)
+		0 = RX FIFO is less than full (or not active TA = 0).
+		1 = RX FIFO is or more full. Cleared by reading sufficient data from the RX FIFO or setting TA to	0.
+
+SPI_CS_TXD		0x00040000
+	TXD TX FIFO can accept Data
+		0 = TX FIFO is full and so cannot accept more data.
+		1 = TX FIFO has space for at least 1 byte.
+
+SPI_CS_RXD		0x00020000
+	RXD RX FIFO contains Data
+		0 = RX FIFO is empty.
+		1 = RX FIFO contains at least 1 byte.
+*/
+static int hm2_rpspi_write(hm2_lowlevel_io_t *llio, uint32_t addr, void *buffer, int size) {
+	//rtapi_print_msg(RTAPI_MSG_ERR, "hm2_rpspi_write addr: %08x, size: %d", addr, size);
+    hm2_rpspi_t *this = (hm2_rpspi_t*) llio;
+    if(size == 0) return 0;
+    if(size % 4) return -EINVAL;
+
+    int msgsize = size/4;
+	uint8_t *tbuff;
+    uint8_t *buffer8 = (uint8_t *)buffer;
+	uint32_t *buffer32 = (uint32_t *)buffer;
+    int i=0, j=0;
+	uint8_t  gully;
+
+    this->txBuf[0] = mk_write_cmd(addr, msgsize, true);
+
+	 for(i=0; i<msgsize; i++)
+		this->txBuf[i+1] = buffer32[i];
+	
+	/* activate transfer and clear fifos */
+	//BCM2835_SPICS = SPI_CS_TA|SPI_CS_CPHA|SPI_CS_CLEAR_TX|SPI_CS_CLEAR_RX;
+	BCM2835_SPICS |= SPI_CS_TA|SPI_CS_CLEAR_TX|SPI_CS_CLEAR_RX;
+
+	tbuff = (uint8_t *)this->txBuf;
+
+	/* send txBuf */
+	/* RPi SPI transfers 8 bits at a time only */
+	for (i=0; i<msgsize+1; i++) {
+		for(j=3; j>=0; j--){
+			BCM2835_SPIFIFO = tbuff[j];
+			if(!(BCM2835_SPICS & SPI_CS_TXD))
+				rtapi_print_msg(RTAPI_MSG_ERR, "TX-FIFO is full!\n");
+		}
+		/* wait until transfer is finished */
+		while (!(BCM2835_SPICS & SPI_CS_DONE));
+		
+		if((BCM2835_SPICS & SPI_CS_RXR) || (BCM2835_SPICS & SPI_CS_RXF))
+			rtapi_print_msg(RTAPI_MSG_ERR, "RX-FIFO is full!\n");
+		
+		for(j=3; j>=0; j--){
+			if(!(BCM2835_SPICS & SPI_CS_RXD))
+				rtapi_print_msg(RTAPI_MSG_ERR, "RX-FIFO is empty!\n");
+			gully = BCM2835_SPIFIFO;
```

Comment:
Perhaps the variable name "gully" has a meaning for you that I don't understand.  

> a water-worn ravine.
> synonyms:	ravine, canyon, gorge, pass, defile, couloir, gulch, coulee, draw
> "a steep icy gully"

I investigated several alternatives in a test program:

~~~~
/*

Tested with gcc 4.9.2 -m32 -O3 -Wall -Wextra -pedantic (debian jessie x86)
and clang 3.5.0-10

All alternatives assemble (on x86) to the sequence
    movl    dev, %eax      (movq dev(%rip),%rax on amd64)
    movzbl  (%eax), %eax   (movzbl (%rax), %eax on amd64)
    ret

Variants fn, gn2, and hn give no diagnotics under gcc.

No variants give any warning under clang (surprising to me)

scan-build gcc warns about variant gn0.

clang++ (note switch to c++ compiler) warns about variants fn and hn:
    warning: expression result unused; assign into a variable to force a volatile load [-Wunused-volatile-lvalue]

g++ gives the same diagnostics as gcc.
*/

#include <stdint.h>

static volatile uint8_t *dev;

void fn() {
    *dev; // discard one byte from read FIFO
}

// gcc issues diagnostic:
//    warning: variable ‘gully’ set but not used [-Wunused-but-set-variable]
//
// scan-build gcc issues diagnostic:
//    Value stored to "gully" is never read
void gn0() {
    uint8_t gully;
    gully = *dev; // discard one byte from read FIFO
}

// gcc issues diagnostics:
//    warning: variable ‘gully’ set but not used [-Wunused-but-set-variable]
//    warning: ignoring #pragma unused  [-Wunknown-pragmas]
void gn1() {
    uint8_t gully;
#pragma unused (gully)
    gully = *dev;
}

void gn2() {
    uint8_t gully __attribute__((unused));
    gully = *dev; // discard one byte from read FIFO
}

void hn() {
    (void)*dev; // discard one byte from read FIFO
}
~~~~

So given the output of all compilers, I think variant gn2 is the best.

No matter which alternative is chosen, **I would greatly prefer to see a comment about the purpose of the line next to the line**, both for myself and for other authors who may not understand this from the variable name.